Runners API

lamdaV opened the pull request:
https://docs.gitlab.com/ce/api/runners.html

Add Missing Runners API.

Not sure how to add tests for this since Runners are another external system to interact with. Mostly manually tested.

gmessner commented on src/main/java/org/gitlab4j/api/models/RunnerAccessLevel.java:
```diff
@@ -0,0 +1,16 @@
+    public String toValue() {
+        return this.value;
+    }
+}
```
Because this is not a shared enum it should be in the RunnersDetail class and follow the convention used elsewhere by using the JacksonJsonEnumHelper:
```
    /** Enum to use for RunnerDetail accessLevel property. */
    public static enum RunnerAccessLevel {

        NOT_PROTECTED, REF_PROTECTED;
        private static JacksonJsonEnumHelper<RunnerAccessLevel > enumHelper = 
            new JacksonJsonEnumHelper<>(RunnerAccessLevel .class);

        @JsonCreator
        public static RunnerAccessLevel forValue(String value) {
            return enumHelper.forValue(value);
        }

        @JsonValue
        public String toValue() {
            return (enumHelper.toString(this));
        }

        @Override
        public String toString() {
            return (enumHelper.toString(this));
        }
    }
```
Using JacksonJsonEnumHelper allows these enums to be consumed without any special logic in the consumer, like looking for null value.

gmessner commented:
BTW, wanted to thank you for this contribution, you've done a great job doing this and following the patterns of the project.

lamdaV commented:
No problem. Thank you for the timely feedback and reviews. This Api wrapper is working nicely with a project I am working on.

lamdaV commented:
Rebase my work onto latest master changes.

gmessner commented:
@lamdaV 
Did you see my comments concerning the 'status' and 'accessLevel' properties?  These should be enums not Strings.

Also, it looks like the comments for RunnerScope and RunnerAccessLevel reversed.

lamdaV commented:
@gmessner 
Thanks for catching that. 

I want to get your thoughts on the naming of `RunnerScope`. I use this for both the typing of `status` for `Runner` and a parameter to take because they are identical in value from my observations from manual testing and reading the docs. Should I rename it to `RunnerStatus` or keep it `RunnerScope`. I originally named it `RunnerScope` because it was related to the `scope` parameter for the Runners Api; however, it is now tied with `status` in the response.

gmessner commented:
@lamdaV 
FYI, Another PR was merged that caused a conflict,  I have resolved it, so you'll need to pull down the change. Was limited to a small change in GitLabApi.java

lamdaV commented:
@gmessner 
Thanks for the reminder. Is this good to go? Let me know if anything is missing or should be changed.

gmessner commented:
@lamdaV 
There are 3 comments above, 2 in the RunnersApi.java and 1 in Runner.java that are pending response.

lamdaV commented:
@gmessner  
I am not seeing some of the comments. Has the review been submitted?

I have addressed the comment asking to use enums helpers for `RunnerAccessLevel` and `RunnerScope` (renamed to `RunnerStatus`) and moved them within the class `RunnerDetail` and `Runner` respectively. 

I changed the `status` and `accessLevel` field to refer to their enums of `RunnerStatus` and `RunnerAccessLevel`.

I did not see any comments in `RunnersApi.java`. Could you check again if the issue still remains?

lamdaV commented on src/main/java/org/gitlab4j/api/RunnersApi.java:
```diff
@@ -0,0 +1,314 @@
+        Response response = delete(Response.Status.OK, formData.asMap(), "projects", projectId, "runners");
+        return (response.readEntity(Runner.class));
+    }
+}
```
This should be addressed now. All method parameters have been more explicit: `runnerId` or `projectId`.

lamdaV commented on src/main/java/org/gitlab4j/api/models/Runner.java:
```diff
@@ -13,6 +17,32 @@
+     */
+    public enum RunnerStatus {
+        SPECIFIC, SHARED, ACTIVE, ONLINE, PAUSED;
+        private static JacksonJsonEnumHelper<RunnerStatus> enumHelper =
```
I have added this. Thanks for catching this.

lamdaV commented on src/main/java/org/gitlab4j/api/RunnersApi.java:
```diff
@@ -0,0 +1,314 @@
+     */
+    public void removeRunner(Integer id) throws GitLabApiException {
+        Response response = delete(Response.Status.NO_CONTENT, null, "runners", id);
+    }
```
I have added null checks for required fields to all methods.

lamdaV commented on src/main/java/org/gitlab4j/api/models/Runner.java:
```diff
@@ -13,6 +13,8 @@
     private Boolean is_shared;
     private String name;
+    private Boolean online;
+    private String status;
```
This has been addressed.

lamdaV commented on src/main/java/org/gitlab4j/api/RunnersApi.java:
```diff
@@ -0,0 +1,316 @@
+     */
+    public Pager<Runner> getRunners(RunnerScope scope, int itemsPerPage) throws GitLabApiException {
+        GitLabApiForm formData = new GitLabApiForm()
+                .withParam("scope", (scope == null) ? null : scope.toValue());
```
All subsequent calls have been addressed. Thanks for informing me of the helper class.

lamdaV commented:
@gmessner 
All comments should now be addressed. Let me know if I missed anything.

gmessner commented:
Changes look great and then some. Merging now and will follow shortly after with an official release.

gmessner commented:
@lamdaV 
GitLab4J-API 4.8.7 has been released with the RunnersApi and ProtectedBranchesApi additions.  Thanks again for your contribution.
